lib/trino-filesystem/src/main/java/io/trino/filesystem/memory/MemoryFileSystemCache.java

```diff
@@ -0,0 +1,219 @@
+        }
+
+        return new MemoryInput(delegate.location(), cachedEntry.get());
+    }
+
+    @Override
+    public TrinoInputStream cacheStream(TrinoInputFile delegate, String key)
+            throws IOException
+    {
+        Optional<Slice> cachedEntry = getFromCache(key, delegate);
+        if (cachedEntry.isEmpty()) {
+            largeFileSkippedCount.incrementAndGet();
+            Span span = createSpan(key, delegate.location(), "delegateStream");
+            return withTracing(span, delegate::newStream);
+        }
+
+        return new MemoryInputStream(delegate.location(), cachedEntry.get());
+    }
+
+    @Override
+    public long cacheLength(TrinoInputFile delegate, String key)
+            throws IOException
+    {
+        Optional<Slice> cachedEntry = getFromCache(key, delegate);
+        if (cachedEntry.isEmpty()) {
+            largeFileSkippedCount.incrementAndGet();
+            Span span = createSpan(key, delegate.location(), "delegateLength");
+            return withTracing(span, delegate::length);
+        }
+
+        return cachedEntry.get().length();
+    }
+
+    @Override
+    public void expire(Location location)
+            throws IOException
+    {
+        List<String> expired = cache.asMap().keySet().stream()
+                .filter(key -> key.contains(location.path()))
+                .collect(toImmutableList());
+        cache.invalidateAll(expired);
+    }
+
+    @Managed
+    public void flushCache()
+    {
+        cache.invalidateAll();
+    }
+
+    @Managed
+    public long getHitCount()
+    {
+        return cache.stats().hitCount();
+    }
+
+    @Managed
+    public long getRequestCount()
+    {
+        return cache.stats().requestCount();
+    }
+
+    @Managed
+    public long getLargeFileSkippedCount()
+    {
+        return largeFileSkippedCount.get();
+    }
+
+    @VisibleForTesting
+    boolean isCached(String key)
+    {
+        Optional<Slice> cachedEntry = cache.getIfPresent(key);
+        return cachedEntry != null && cachedEntry.isPresent();
+    }
+
+    private Optional<Slice> getFromCache(String key, TrinoInputFile delegate)
+            throws IOException
+    {
+        try {
+            // File length is available for free at this point, we add it to cache key to be resilient to manual modifications to file
+            return cache.get(key, () -> load(key, delegate));
+        }
+        catch (ExecutionException e) {
+            throw handleException(delegate.location(), e.getCause());
+        }
+    }
+
+    private Optional<Slice> load(String key, TrinoInputFile delegate)
+            throws IOException
+    {
+        Span span = createSpan(key, delegate.location(), "loadCache");
+        return withTracing(span, () -> {
+            long fileSize = delegate.length();
+            span.setAttribute(CACHE_FILE_READ_SIZE, fileSize);
+            if (fileSize > maxContentLengthBytes) {
+                return Optional.empty();
+            }
```

Comment:
What if we want to extend this kind of caching to Delta Lake ?
There we have some metadata files (the stats & _last_checkpoint) which are bound to change and are not to be cached. Pls think of a mechanism that we could get around in such situations.

Reply:
Nothing special is needed for Delta lake, DeltaLakeCacheKeyProvider will take care of this